module/zfs/range_tree.c

```diff
@@ -318,14 +339,25 @@ zfs_range_tree_add_impl(void *arg, uint64_t start, uint64_t size, uint64_t fill)
 	 * the normal code paths.
 	 */
 	if (rs != NULL) {
+		uint64_t rstart = zfs_rs_get_start(rs, rt);
+		uint64_t rend = zfs_rs_get_end(rs, rt);
 		if (gap == 0) {
-			zfs_panic_recover("zfs: adding existent segment to "
-			    "range tree (offset=%llx size=%llx)",
-			    (longlong_t)start, (longlong_t)size);
+			zfs_panic_recover_ms("zfs: adding segment "
+			    "(offset=%llx size=%llx) overlapping with "
+			    "existing one (offset=%llx size=%llx)",
```

Comment:
Wish we could say something about what was the range tree and what we were doing, otherwise just a message that something on pool overlapped somewhere does not help us with debugging.  In case of panic we should get a stack, but I wonder if we could get more.  May be giving range trees some types, names, etc. for debugging.

Reply:
Thinking of more generalized approach, I guess we could cover both comments with something like `zfs_recover_range_tree_mask` (instead of `zfs_recover_ms`) which allows selecting up to 64 most interesting range tree classes: `ms_allocatable`, `ms_freeing`, `ms_freed`, and so on. I think 64 bits are enough to cover all range tree use cases we have currently, even the ones not related to spacemaps. Each instance of the selected tree classes would go the warning path instead of the panic one. In addition, each range tree instance could really have some description which says about the actual name like `ms_allocatable` and, probably, extra info like spa/vdev/metaslab id or something. Does it seem like one of the options we could consider implementing?

Reply:
I think what @amotin had in mind was just the one tunable, but make the log output mention which range-tree, and as much details as possible so we can have something to track down when a user reports encountering the problem. Knowing which of the range trees would be a good first step, but it'd also be nice to show the range that was being added/removed, the full details of the range it overlapped with (do we have a birth time for each available?), and in general to just leave as many bread crumbs as possible to chase this problem down

Reply:
Yes, it seems that the range tree "selector" using some mask is an unnecessary extra complexity.

Reply:
I was thinking about range trees having some names, similar to different kernel locks on FreeBSD, that allow to understand the context when something goes wrong.  I haven't checked, but I don't think we have so many range trees that additional const char pointer would matter.  But as Allan told, we should try to log as much as we can.